jax/lax/lax_control_flow.py

```diff
@@ -974,14 +977,141 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
+                      transpose_solve, symmetric):
+  # A x - b = 0
+  # ∂A x + A ∂x - ∂b = 0
+  # ∂x = A^{-1} (∂b - ∂A x)
+
+  solve_kwargs = dict(num_consts=num_consts, jaxpr=jaxpr, solve=solve,
+                      transpose_solve=transpose_solve, symmetric=symmetric)
+
+  x = linear_solve_p.bind(*primals, **solve_kwargs)
+
+  params, _ = split_list(primals, [num_consts])
+  params_dot, b_dot = split_list(tangents, [num_consts])
+
+  matvec = partial(core.jaxpr_as_fun(jaxpr), *params)
+  matvec_dot = partial(
+      _tangent_linear_map, core.jaxpr_as_fun(jaxpr), params, params_dot
+  )
+
+  if all(tangent is ad_util.zero for tangent in params_dot):
+    rhs = b_dot
+  elif all(tangent is ad_util.zero for tangent in b_dot):
+    rhs = [-u for u in matvec_dot(*x)]
+  else:
+    rhs = [u - v for u, v in zip(b_dot, matvec_dot(*x))]
+
+  x_dot = linear_solve_p.bind(*(params + rhs), **solve_kwargs)
+  return x, x_dot
+
+
+def _transpose_function(linear_fun, xs):
+  """Transpose a linear function."""
+  # TODO(shoyer): can we use something more direct than the vjp machinery?
+  # It's particularly awkward that we need the second argument to give
+  # particular values of the primals, which are entirely arbitrary.
+  _, transposed_fun = ad.vjp(lu.wrap_init(linear_fun), xs)
+  return transposed_fun
+
+
+def _linear_solve_transpose_rule(cotangent, *primals, **kwargs):
+  num_consts, jaxpr, solve, transpose_solve, symmetric = split_dict(
+      kwargs, ['num_consts', 'jaxpr', 'solve', 'transpose_solve', 'symmetric'])
+  params, b = split_list(primals, [num_consts])
+  assert b == [ad.undefined_primal] * len(b)
+
+  if not symmetric:
+    # TODO(shoyer): make this less awkward. Maybe we should compute this jaxpr
+    # for vector-matrix products once before calling the primitive?
+    vecmat = _transpose_function(
+        partial(core.jaxpr_as_fun(jaxpr), *params), cotangent)
+    # We can throw away the (re)flattened cotangents and output pytree, since we
+    # know cotangent and the outputs of vecmat are already a flattened.
+    _, in_tree = tree_flatten(tuple(cotangent))
+    avals = tuple(_map(_abstractify, cotangent))
+    jaxpr, params, _ = _initial_style_jaxpr(vecmat, in_tree, avals)
+    params = list(params)
+
+  cotangent_b = linear_solve_p.bind(
+      *(params + cotangent), num_consts=num_consts, jaxpr=jaxpr,
+      solve=transpose_solve, transpose_solve=solve, symmetric=symmetric)
+
+  return [None] * num_consts + cotangent_b
+
+
+linear_solve_p = core.Primitive('linear_solve')
+linear_solve_p.multiple_results = True
+linear_solve_p.def_impl(_linear_solve_impl)
+linear_solve_p.def_abstract_eval(_linear_solve_abstract_eval)
+ad.primitive_jvps[linear_solve_p] = _linear_solve_jvp
+xla.initial_style_translations[linear_solve_p] = xla.lower_fun(
+    _linear_solve_impl, initial_style=True)
+ad.primitive_transposes[linear_solve_p] = _linear_solve_transpose_rule
+batching.deftraced(linear_solve_p)
```

Comment:
Do you have a sense of whether this will run into #1249?

Reply:
Yeah, I was wondering the same thing...

Reply:
Yeah, I was wondering the same thing...

Reply:
Indeed, this most definitely suffers from the same issue. It appears we will need to write custom batching rules for `root` and `linear_solve`. I think this would be pretty doable, but now I'm leaving this as a "TODO".